File: api/src/api/services/hadith_translation.py
"""Indonesian translations for hadith corpora.

The seeded kitab corpus has Indonesian text only for the Qur'an
(Kemenag). Hadith corpora (Bukhari, Muslim, Riyad as-Salihin) carry
Arabic + English only — which meant every hadith-cited flyer rendered
in English on the Bahasa-default UI.

This service translates English hadith text into Indonesian via
Gemini Flash-Lite, caching per `(corpus, hadithnumber)` in
`hadith_translations_id`. First call for a given hadith pays the
LLM cost; subsequent calls are free SELECT.

Sharia-aware prompt: faithful translation, no interpretation,
preserves the speaker, the chain ("Rasulullah bersabda…"), and the
specific words. Renders English idioms into natural Indonesian without
adding commentary.

Cost: ~Rp 5–10 per hadith on Flash-Lite. With ~50–100 distinct hadith
cited across briefings over a quarter, lifetime cost is trivial.
"""

from __future__ import annotations

import asyncio
import time

import structlog
from google import genai
from google.genai import errors as genai_errors
from google.genai import types
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.ext.asyncio import AsyncSession

from api.config import settings
from api.models.admin import HadithTranslationId

log = structlog.get_logger(__name__)

MODEL = "gemini-2.5-flash-lite"
MAX_RETRIES = 3
RETRY_BASE_SLEEP_S = 4.0

SYSTEM_PROMPT = """Anda penerjemah teks hadits dari Bahasa Inggris ke \
Bahasa Indonesia.

Aturan:
- Terjemahkan dengan setia. Jangan tambah tafsir, jangan kurangi makna, \
jangan ringkas.
- Pertahankan nama-nama (Nabi, sahabat, perawi) seperti aslinya.
- Sebutan untuk Rasulullah ﷺ tetap "Rasulullah" atau "Nabi" sesuai \
teks sumber. Jangan ditambah salawat sendiri.
- Frasa yang lazim dalam tradisi hadits ID gunakan bentuk standar: \
"bersabda" (said), "diriwayatkan oleh" (narrated by), \
"radhiyallahu 'anhu" (may Allah be pleased with him), dst.
- Output: hanya hasil terjemahan dalam Bahasa Indonesia. \
Tanpa preamble, tanpa label, tanpa kutip pembuka/penutup.
- Jaga gaya Bahasa Indonesia yang natural dan mudah dibaca oleh \
audiens umum (santri, mahasiswa, pengurus pengajian)."""

_client: genai.Client | None = None


def _get_client() -> genai.Client:
    if not settings.gemini_api_key:
        raise RuntimeError("GEMINI_API_KEY is not set. Add to .env.")
    global _client
    if _client is None:
        _client = genai.Client(api_key=settings.gemini_api_key)
    return _client


def _call_gemini(text_en: str) -> str:
    """Synchronous Gemini call wrapped for the asyncio thread offload."""
    client = _get_client()
    last_exc: Exception | None = None
    for attempt in range(MAX_RETRIES):
        try:
            resp = client.models.generate_content(
                model=MODEL,
                contents=text_en,
                config=types.GenerateContentConfig(
                    system_instruction=SYSTEM_PROMPT,
                    temperature=0.2,
                ),
            )
            break
        except genai_errors.ServerError as exc:
            last_exc = exc
            if attempt == MAX_RETRIES - 1:
                raise
            time.sleep(RETRY_BASE_SLEEP_S * (2**attempt))
    else:
        raise RuntimeError("translate_hadith: retry loop exited") from last_exc

    out = (resp.text or "").strip()
    if not out:
        raise RuntimeError("translate_hadith: empty model response")

    from api.services.usage import gemini_output_tokens, record_usage

    usage_md = getattr(resp, "usage_metadata", None)
    record_usage(
        provider="gemini",
        operation="translate_hadith_to_id",
        model=MODEL,
        tokens_in=getattr(usage_md, "prompt_token_count", None),
        tokens_out=gemini_output_tokens(usage_md),
        meta={"chars_in": len(text_en), "chars_out": len(out)},
    )
    return out


async def translate_hadith_to_id(
    session: AsyncSession,
    corpus: str,
    hadithnumber: str,
    text_en: str,
) -> str:
    """Return the Indonesian translation of `text_en`, using the
    `hadith_translations_id` cache. On a cache miss, calls Gemini
    Flash-Lite and persists the result. Empty `text_en` short-circuits
    to "" so callers don't need to pre-check."""
    if not text_en.strip():
        return ""

    # Cache hit — most calls. Re-check the English source so a corpus
    # refresh that altered the upstream text re-triggers translation.
    row = await session.get(HadithTranslationId, (corpus, hadithnumber))
    if row is not None and row.text_en == text_en:
        return row.text_id

    # Cache miss / source-changed. Translate, then upsert.
    try:
        text_id = await asyncio.to_thread(_call_gemini, text_en)
    except Exception as exc:
        log.warning(
            "hadith_translation.translate_failed",
            corpus=corpus,
            hadithnumber=hadithnumber,
            error=str(exc)[:200],
        )
        # Don't poison the cache with a bad translation — return ""
        # so the flyer renderer falls back to English for this cycle.
        return ""

    stmt = (
        pg_insert(HadithTranslationId)
        .values(
            corpus=corpus,
            hadithnumber=hadithnumber,
            text_en=text_en,
            text_id=text_id,
            model=MODEL,
        )
        .on_conflict_do_update(
            index_elements=["corpus", "hadithnumber"],
            set_={
                "text_en": text_en,
                "text_id": text_id,
                "model": MODEL,
            },
        )
    )
    await session.execute(stmt)
    await session.commit()
    return text_id


async def enrich_daleel_translations(
    session: AsyncSession,
    hits: list[dict],
) -> list[dict]:
    """Fill `translation_id` for hadith hits whose `translation_id` is
    blank. Mutates and returns the list. Qur'an hits are untouched —
    they already carry Kemenag ID text from the corpus."""
    for hit in hits:
        if hit.get("corpus") == "quran":
            continue
        if hit.get("translation_id"):
            continue
        text_en = (hit.get("translation_en") or "").strip()
        if not text_en:
            continue
        ref_id = hit.get("ref_id", "")
        # ref_id is `{corpus}::{hadithnumber}` — split back out so the
        # cache key matches what's stored in the table.
        parts = ref_id.split("::", 1)
        if len(parts) != 2:
            continue
        corpus, hadithnumber = parts
        if not hadithnumber:
            continue
        translated = await translate_hadith_to_id(
            session, corpus, hadithnumber, text_en
        )
        if translated:
            hit["translation_id"] = translated
    return hits

File: web/src/lib/hadith-translation.ts
import "server-only";
import { and, eq } from "drizzle-orm";
import { GoogleGenAI } from "@google/genai";

import { db, schema } from "@/db";
import { recordUsage } from "@/lib/usage-log";

/**
 * Lookup the Indonesian translation of a hadith from the
 * `hadith_translations_id` cache. On a cache miss, call Flash-Lite to
 * translate, persist to the cache, then return.
 *
 * Mirrors the Python helper `services/hadith_translation.py` so the
 * cache stays consistent across both runtimes. Same model
 * (gemini-2.5-flash-lite), same prompt shape — a row written from
 * either side is readable by both.
 *
 * Returns "" on failure (LLM error, no key) so the caller falls back
 * to English without breaking.
 */

const MODEL = "gemini-2.5-flash-lite";

// Map the short KitabCorpus enum the web uses (riyad, bulugh) to the
// FULL slug Python writes into the cache (riyad_as_salihin,
// bulugh_al_maram). Misses → return null → caller skips translation.
const CORPUS_SLUG: Record<string, string> = {
  bukhari: "bukhari",
  muslim: "muslim",
  riyad: "riyad_as_salihin",
  bulugh: "bulugh_al_maram",
};

let _client: GoogleGenAI | null = null;
function getClient(): GoogleGenAI | null {
  if (!process.env.GEMINI_API_KEY) return null;
  if (_client === null) {
    _client = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  }
  return _client;
}

const PROMPT = (textEn: string) => `Terjemahkan hadits berikut ke Bahasa Indonesia yang baku dan natural. Pertahankan struktur Islami: nama-nama sahabat + greeting (RA, ﷺ) tetap apa adanya, citation di akhir kalau ada juga tetap. JANGAN tambahkan komentar atau penjelasan. Output HANYA terjemahan Indonesia-nya, tanpa pengantar atau penutup.

Teks bahasa Inggris:
"""
${textEn}
"""

Terjemahan Bahasa Indonesia:`;

async function translateViaLlm(textEn: string): Promise<string> {
  const client = getClient();
  if (!client) return "";
  try {
    const resp = await client.models.generateContent({
      model: MODEL,
      contents: PROMPT(textEn),
      config: {
        temperature: 0.2,
        maxOutputTokens: 600,
        thinkingConfig: { thinkingBudget: 0 },
      },
    });
    void recordUsage({
      provider: "gemini",
      operation: "translate_hadith_to_id",
      model: MODEL,
      tokensIn: resp.usageMetadata?.promptTokenCount ?? null,
      tokensOut: resp.usageMetadata?.candidatesTokenCount ?? null,
    });
    return (resp.text ?? "").trim();
  } catch (err) {
    console.warn(
      "[hadith-translation] LLM translate failed:",
      err instanceof Error ? err.message : err,
    );
    return "";
  }
}

/**
 * Returns the Indonesian translation of a hadith, hitting the cache
 * first and falling back to a Flash-Lite translate-then-persist.
 *
 * - `corpus`: short slug as returned by `searchKitabBrowse` (`riyad`,
 *   `bulugh`, `bukhari`, `muslim`). Internally mapped to the full slug
 *   the Python pipeline writes.
 * - `hadithNumber`: numeric or string identifier inside the corpus.
 * - `textEn`: the English text we currently have. If the cached row's
 *   `text_en` doesn't match (upstream corpus refresh), we re-translate.
 */
export async function translateHadithToId({
  corpus,
  hadithNumber,
  textEn,
}: {
  corpus: string;
  hadithNumber: string | number;
  textEn: string;
}): Promise<string> {
  if (!textEn.trim()) return "";
  const slug = CORPUS_SLUG[corpus];
  if (!slug) return "";
  const hadithStr = String(hadithNumber);

  // 1. Cache lookup.
  const [cached] = await db
    .select({
      textEn: schema.hadithTranslationsId.textEn,
      textId: schema.hadithTranslationsId.textId,
    })
    .from(schema.hadithTranslationsId)
    .where(
      and(
        eq(schema.hadithTranslationsId.corpus, slug),
        eq(schema.hadithTranslationsId.hadithnumber, hadithStr),
      ),
    )
    .limit(1);
  if (cached && cached.textEn === textEn) {
    return cached.textId;
  }

  // 2. Cache miss / English changed → translate and upsert.
  const translated = await translateViaLlm(textEn);
  if (!translated) return "";

  await db
    .insert(schema.hadithTranslationsId)
    .values({
      corpus: slug,
      hadithnumber: hadithStr,
      textEn,
      textId: translated,
      model: MODEL,
    })
    .onConflictDoUpdate({
      target: [
        schema.hadithTranslationsId.corpus,
        schema.hadithTranslationsId.hadithnumber,
      ],
      set: {
        textEn,
        textId: translated,
        model: MODEL,
      },
    });

  return translated;
}
